### DISCUSSION TITLE
Combining JAX and xarray for large scale computing

### DISCUSSION THREAD

**Discussion by Aceticia** [pydata/xarray#8164] [Q&A] [answered]
I'm working on computational neuroscience and I do a lot of large scale computations (perhaps not as big as genomics folks). These computations are usually simple linear algebra manipulations, parallelized to a bunch of CPUs. Meanwhile, our lab have a large number of GPUs for use. What would be a good way to take advantage of the easy usage of xarray and JAX?

In my workflow, I already have a lot of vectorized calls and it's saving me a lot of time. But these are still a large number of small linear algebra operations that might see huge gains when performed on GPUs.

A naive way is to simply convert to jax in a vectorized function. For example, I vectorize the calls to a number of brain scans. [num_subjects, num_conditions, x, y, z, features] where xyz is the spatial dimensions. I have a function `func` that processes certain features. I apply ufunc to vectorize over the first 5 axes. Then I just make `func` jax, and don't touch anything else.

But this is likely going to incur a large conversion overhead. In the case that entire data fits into memory un-chunked, what do you think is a better approach?

**Comment by keewis** [comment, marked as answer]
if you repeatedly would be converting to / from jax, it would probably be best to keep everything in jax and have `xarray` wrap that. 

According to google/jax#17107, support for doing so seems to be close but not quite there yet. In particular we'd need jax to implement the array API (i.e., define `__array_namespace__`) – see google/jax#16099 a in-progress PR.

**Comment by ConstantinVasilev** [comment]
Hi @keewis, the Jax PR (https://github.com/google/jax/pull/16099) is merged, so is there an updated answer to this discussion?

**Reply by keewis** [reply]
nothing that I'm aware of, but I'm not using JAX myself so I'm not the best person to ask.

However, since JAX appears to have Array API compatibility now, I'd invite you to just try it out: create a JAX array and wrap it using a `Dataset` or `DataArray` object. If that works, try calling a couple of methods and see what fails, and in any case I'd encourage you to report back here on how it went (also, feel free to ask any questions regarding `xarray` usage here).

**Comment by zazbone** [comment]
Hi !
I'm in a similar situation where I use Jax as computation backend, but I could also benefit from xarray Dataset
I've tried to init a DataArray with as Jax array as input data, but it looks like xarray does an automatic conversion to ndarray
I raise the subject again to know if there is update or plan on jax or array api standard support 

Cheers

**Reply by keewis** [reply]
looks like you have to use `jax.experimental.array_api` instead of `jax.numpy` for this to work:
```python
In [5]: import jax.experimental.array_api as jxp
   ...: import xarray as xr
   ...: 
   ...: arr = xr.DataArray(jxp.zeros((5, 3)), dims=["y", "x"])
   ...: arr.sum(dim="y")
Out[5]: 
<xarray.DataArray (x: 3)> Size: 12B
Array([0., 0., 0.], dtype=float32)
Dimensions without coordinates: x
```
(note the `Array` instead of `array`)

which doesn't mean _all_ `xarray` operations are supported (see [the list of known issues](https://docs.xarray.dev/en/latest/user-guide/duckarrays.html#conversion-to-numpy-as-a-fallback), and the rest you'll have to check yourself), but at least standard reductions appear to have worked fine.

**Reply by TomNicholas** [reply]
Wait this is big! I didn't know `jax.experimental.array_api` existed!

Now would be a really good time to finish off that framework for automatically testing different duck-typed arrays in xarray... 😅

**Reply by keewis** [reply]
that will take a lot of concentrated effort... we can try doing some kind of on-call sprint, if you'd like (if so, to be discussed somewhere else).

**Reply by TomNicholas** [reply]
I would be down for that actually. I also met a bunch of people at SciPy who are interested in ragged arrays in xarray. But I am supposed to be on holiday right now, so maybe remind me sometime after next week? 😁

**Reply by zazbone** [reply]
Ho, good to see it work, I will wait until Jax stabilized it but it's cool
Thanks

**Comment by allen-adastra** [comment]
Starting from JAX `0.4.32`, we **should** be able to wrap `jax.Array` with xarray structures.

Two additional issues for Xarray + JAX integration have been identified:
    1. Automatic dispatching based on array type https://github.com/pydata/xarray/issues/7848
    2. Registering Xarray structures as PyTree nodes

Issue 2 has an example solution in the [GraphCast project](https://github.com/google-deepmind/graphcast/blob/main/graphcast/xarray_jax.py).

The main challenge I've run into there is that many JAX libraries (e.g. [Diffrax](https://github.com/patrick-kidger/diffrax)) use tree manipulation patterns that produce "dummy trees" like boolean masks. If a `xr.DataArray` has its `data` attribute replaced with a `True`, it will trigger all kinds of errors.

I can think of three solutions to this:
1. Avoid inputting xarray types into functions that do this kind of masking
2. Introduce flags into xarray to disable checks to allow for things like tree masks
3. Implement flatten functions that convert xarray types into another type (e.g. `XJDataArray`) and implement `to_xarray()` methods to allow the user to get back the original structure inside jitted functions ([experimental implementation of this approach](https://github.com/allen-adastra/xarray_jax))

**Reply by mjwillson** [reply]
Yeah I've run into this problem before too and not sure if there's a super elegant solution.

Another possible solution: in the unflatten function, if you see that the leaves are obviously not xarray-compatible values (e.g. None, True, False etc), generate stub objects that are also registered as PyTree nodes, e.g.

```python
@dataclasses.dataclass(frozen=True)
class DataArrayLikeGenericTreeNode[_T]:
  data: _T
```

which 'quacks like' a DataArray in as much as is necessary for it to behave the same way in a generic pytree context. (Since libraries that are producing these 'dummy trees' will generally be agnostic to the specific node types of the tree, or certainly not relying on xarray functionality on them at least.)

If you then mapped one of these generic tree node classes back to jax arrays again using another jax.tree.map or similar, you'd probably want it to unflatten to an actual DataArray, but that could be arranged too.

Could still be a problem for any code that expects the types of the nodes to match (DataArray vs DataArrayLikeGenericTreeNode). I suspect it's not common to rely on this or check for it.

Overall still feels kinda hacky but might be the least intrusive on users?

**Reply by mjwillson** [reply]
Do you have an example of a tree manipulation pattern that triggers an issue at the moment? (Ideally fairly minimal / self-contained).

I realise the distinction of 'obviously not xarray-compatible values (e.g. None, True, False etc)' is a bit vague, since you might conceivably want True/False to be converted to a boolean scalar jax array, and what about other python scalars etc. But I think in practice it might work well enough to assume anything that's not either a jax array, numpy array or (perhaps?) a numeric python scalar would trigger the dummy tree node class to be used.

**Reply by allen-adastra** [reply]
> Do you have an example of a tree manipulation pattern that triggers an issue at the moment? (Ideally fairly minimal / self-contained).
> 
> I realise the distinction of 'obviously not xarray-compatible values (e.g. None, True, False etc)' is a bit vague, since you might conceivably want True/False to be converted to a boolean scalar jax array, and what about other python scalars etc. But I think in practice it might work well enough to assume anything that's not either a jax array, numpy array or (perhaps?) a numeric python scalar would trigger the dummy tree node class to be used.

Posted a notebook example that boils down to creating a tree mask:
```
jax.tree.map(lambda x: x.ndim == 2, ds)
```

https://github.com/allen-adastra/xarray_jax/blob/allenw/minimal_tree_manip_example/tree_manip_issue.ipynb

**Reply by allen-adastra** [reply]
I'm a bit concerned that using a dummy PyTree node will cause problems because now the PyTree technically has a different structure. Won't that create issues with things like multi-mapping?
```
jax.tree.map(some_fun, tree, tree_mask)
```
FWIW, when I went into Xarray and manually disabled checks for the input array, I ended up getting things to work. But that will require the Xarray devs to commit to supporting a `disable_checks` feature or something like that. Curious what @TomNicholas thinks?

**Reply by mjwillson** [reply]
Dammit, you're right this won't work with multi-argument `jax.tree.map`.

I have one further idea, also a total hack but maybe it will appeal. The trick is that you don't actually have to construct the xarray.Variable (or whatever) with the None, True/False or whatever as the data. You merely have to write unflatten/flatten functions that makes it look like you did to jax.tree_util. Because the actual underlying custom datatype is opaque to jax.tree_util and is only made accessible to it via your flatten/unflatten function.

Furthermore you can use xarray attributes to store extra side-channel information about what the data structure is really supposed to look like. So you can do something like this:

```python
import xarray
import jax

def flatten_variable(v):
  if '_original_dims' in v.attrs:
    children = (v.data.item(),)
    aux = v.attrs['_original_dims']
  else:
    children = (v.data,)
    aux = v.dims
  return children, aux

def unflatten_variable(aux, children):
  data = children[0]
  dims = aux
  if data in (True, False, None):
    return xarray.Variable((), data, attrs={'_original_dims': dims})
  else:
    return xarray.Variable(dims, data)

jax.tree_util.register_pytree_node(xarray.Variable, flatten_variable, unflatten_variable)
```

Then things like this work:

```python
v = xarray.Variable(('x',), [1.23])
v_bool = jax.tree.map(lambda x: True, v)
jax.tree.map(lambda x, flag: x+1 if flag else x-1,
             v,
             v_bool)
```

Note I had to make sure the 'aux' always corresponds to the original set of dimensions, because jax will complain if it doesn't match.

Getting something similar to work for DataArray and Dataset may be more fiddly, because xarray would require the shape of the data to match the coordinates etc. For that, you might have to pass in some dummy data (e.g. a lazy dask array of zeros to avoid overhead) with the right shape, and store the actual True/False/None value in an attribute. (Or I guess you could have a lazy dask array that repeats that value, if you can avoid realising the full thing and still extract the value). Gonna leave that as an exercise for the reader, for now anyway :)

**Reply by mjwillson** [reply]
The root problem in all this is that jax.tree_util assumes that tree nodes belong to a generic data type that can hold anything, but that's just not true in general and IMO not reasonable to expect for custom node types.

A possible solution in jax.tree_util for this could be this:
* When registering a node type, you specify if it's generic or not
* When unflattening a non-generic node type, if the type of the value doesn't match the original type, instead it falls back on representing the entire resulting tree as a special class which just wraps the pair of (treedef, leaf nodes). This custom class is then supported by all the rest of the API. Eventually if it's mapped / unflattened back with the original types matching up, it unflattens properly to the original node type.

Kind of complicated to implement, though, and people would have to be warned to expect if registering a non-generic node type.

Or, one can limit ones expectations around how much generic tree manipulation will be possible with non-generic node types. That's essentially the approach I've taken so far.

Another approach could be that libraries that want to do clever stuff with filtering nodes etc, use the "pair of (treedef, leaf nodes)" representation internally for their trees-of-booleans etc, and don't try to unflatten them.

**Reply by allen-adastra** [reply]
Hmm... yeah this is a doozy.

I've personally found things like [`eqx.partition`](https://github.com/patrick-kidger/equinox/blob/fab726bdd15dd95840088427dcc5fad58fd8d435/equinox/_filters.py#L135) to be extremely useful, and have a bit of a personal dependency on `eqx` and `diffrax`, so I'm inclined to go with a solution that preserves compatibility with those two libraries.

When I have more time (probably on a weekend D:), I can fiddle around with your `attrs` solution in the `DataArray` and `Dataset` case.

Failing that, I might just go with whats in the [experimental repo](https://github.com/allen-adastra/xarray_jax) for now and suffer the pain of manually calling `.to_xarray()` periodically.

**Reply by TomNicholas** [reply]
> when I went into Xarray and manually disabled checks for the input array, I ended up getting things to work. But that will require the Xarray devs to commit to supporting a disable_checks feature or something like that.

Can you elaborate on this @allen-adastra ? Which checks exactly have to be bypassed? There may be a relatively non-invasive way of allowing this.

I also would like to know if @shoyer (or anyone from his team) has thoughts about the ideal way to do this Xarray-JAX integration.

**Reply by allen-adastra** [reply]
The fundamental issue at hand is:  currently, Xarray types require `data` attributes that are consistent with dimensions/coords, but to do common JAX tree manipulation things, we need to allow dummy `data` attributes.

In `xr.Variable` and `xr.DataArray` there are checks that the wrapped data is the same dim as the dimensions. After I comment out these checks, things work as expected.

`xr.Dataset` is a bit more complicated. I have more of an explanation in [this notebook](https://github.com/allen-adastra/xarray_jax/blob/allenw/disable_xarray_checks/mask_examples.ipynb).

On [that branch](https://github.com/allen-adastra/xarray_jax/tree/allenw/disable_xarray_checks), I have a fork of `xarray` as a submodule, and in the pinned commit you can search for `# Disabled for xarray_jax experiment.` to find the lines of code I have disabled to get `xr.Variable` and `xr.DataArray` working.

**Reply by shoyer** [reply]
I wonder if "Potential solution 2" ("Structure your custom `tree_unflatten` function so that it avoids calling `__init__` ") from the [JAX docs](https://jax.readthedocs.io/en/latest/working-with-pytrees.html#custom-pytrees-and-initialization-with-unexpected-values) would suffice here.

(Incidentally, I think we will still want a small `xarray_jax` package to implement dimension-name aware JAX transforms like `vmap` for Xarray objects.)

**Reply by allen-adastra** [reply]
Nice; that works well for `Variable` and `DataArray`, but:
```
ds = object.__new__(xarray.Dataset)
```
Seems to generate a recursion serror

**Reply by shoyer** [reply]
You also have to manually copy attributes if you want `__new__` to work, e.g.,
```python
ds2 = object.__new__(xarray.Dataset)
ds2._dims = ds._dims
ds2._variables = ds._variables
ds2._coord_names = ds._coord_names
ds2._attrs = ds._attrs
ds2._indexes = ds._indexes
ds2._encoding = None
ds2._close = None
```

**Reply by allen-adastra** [reply]
Yep, what I mean is independent of the JAX context, if you do this:
```
import xarray as xr
ds = object.__new__(xarray.Dataset)
print(ds)
```
you get a recursion error.

So I've made solid progress with `xr.Variable` and `xr.DataArray`. I've cut a new branch in my `xarray_jax` repo, and the tests for `xr.Variable` and `xr.DataArray` [here](https://github.com/allen-adastra/xarray_jax/blob/allenw/use_object_new/tests/test_types.py) pass (note: I had to set `xr.set_options(keep_attrs=True)` to preserve PyTree structure).

I'm running into a recursion error at [this spot](https://github.com/allen-adastra/xarray_jax/blob/allenw/use_object_new/tests/test_types.py#L121). If you have ideas/pointers that would be very welcome!

**Reply by allen-adastra** [reply]
Update: I've merged a bunch of changes into `main` of the [`xarray_jax` repo](https://github.com/allen-adastra/xarray_jax) along with more info in the `README.md` on examples, supported features and sharp edges we've found.

Still don't have a solution for registering `xr.Dataset` yet, but for now the following seems to work:

```
ds = xr.tutorial.load_dataset("air_temperature")

@eqx.filter_jit
def some_function(xjds: xj.XjDataset):
    # Convert to xr.Dataset.
    xrds = xj.to_xarray(xjds)

    # Do some operation.
    xrds = -1.0 * xrds

    # Convert back to xj.Dataset.
    return xj.from_xarray(xrds)

xjds = some_function(xj.from_xarray(ds))
ds_new = xj.to_xarray(xjds)
```

**Reply by shoyer** [reply]
@allen-adastra In your failing example with `Dataset`, you are iterating over `Dataset.data_vars`. I'm not sure we can reasonably expect that or any other transformation that creating invalid xarray `data` values to work. What _should_ work are cases where JAX uses pytrees for internal book-keeping but doesn't surface those in external APIs.

I would love to see a specific list of JAX transformations (including external ones like Equinox and NNX) that do or do not work.

**Reply by allen-adastra** [reply]
Yeah sounds good. I've got a [test with some hypothesis strategies for types](https://github.com/allen-adastra/xarray_jax/blob/main/tests/test_types.py), would make sense to setup another one with transforms.

Do we have a `xr.DataArray` or `xr.Dataset` strategy btw?

I'm relatively new to Xarray + Hypothesis as well, so def would  welcome ideas + feedback!

**Reply by allen-adastra** [reply]
Something else I've run into: `xr.IndexVariable` converts `JAX` arrays to numpy arrays. Is this intended? Seems like it makes sense but just want to check.

```
@given(var=index_variables, ufunc=ufunc_strat)
@settings(deadline=None)
def test_index_variables(var: xr.Variable, ufunc):
    var = var.to_index_variable()

    # Index variables should not be wrapped in a jax array.
    assert not isinstance(var.data, jax.Array)
```

**Reply by TomNicholas** [reply]
> I've got a [test with some hypothesis strategies for types](https://github.com/allen-adastra/xarray_jax/blob/main/tests/test_types.py),

Yay someone's actually using that!

> Do we have a `xr.DataArray` or `xr.Dataset` strategy btw?

No, but there is a 90% finished PR to add them... (https://github.com/pydata/xarray/pull/6908)

**Reply by allen-adastra** [reply]
Ah cool! Would be cool to have that (not that you don't already have a lot going on here :>))

**Reply by shoyer** [reply]
> Something else I've run into: `xr.IndexVariable` converts `JAX` arrays to numpy arrays. Is this intended? Seems like it makes sense but just want to check.

Yes, this is intended. Variables with an index get loaded into memory, with an associated `pandas.Index`. They should be treated as static information as far as JAX is concerned.

**Reply by allen-adastra** [reply]
> @allen-adastra In your failing example with `Dataset`, you are iterating over `Dataset.data_vars`. I'm not sure we can reasonably expect that or any other transformation that creating invalid xarray `data` values to work. What _should_ work are cases where JAX uses pytrees for internal book-keeping but doesn't surface those in external APIs.
> 
> I would love to see a specific list of JAX transformations (including external ones like Equinox and NNX) that do or do not work.

Merged a [PR](https://github.com/allen-adastra/xarray_jax/pull/4) that does checks some basic JAX and Equinox transforms, and things seem to work pretty well as long as the shape of the underlying array doesn't change (including `jax.grad`, although you gotta make sure you expose the underlying array with `.data` on the function output).

I saw @mjwillson had a particular solution for handling dim changes in the GraphCast project, but that requires the user to manually specify a context. Would be nice if there was some way to add dummy dimensions automatically.

**Comment by allen-adastra** [comment]
I pushed a [`v0.0.1` solution](https://github.com/allen-adastra/xarray_jax/tree/main) to `PyPi` with an acknowledgement of the folks in this thread. Would welcome testers to find new and exciting ways to break it :)

**Comment by allen-adastra** [comment]
https://github.com/pydata/xarray/blob/44c7c151717a97208e4e7bc069587b48a76a6e61/xarray/core/variable.py#L338 seems to cause issues with Xarray + JAX integration

**Reply by keewis** [reply]
just so I can understand, how do you end up passing a `jax` array to that function? I'd expect anything implementing `__array_namespace__` to never hit that unless we're explicitly converting to `numpy`.
